### PR TITLE
Detect separate git dir and set git config path value appropriately

### LINKED ISSUE #30034: Git module fails to update if separated work tree
##### ISSUE TYPE

Bug Report

##### COMPONENT NAME

git module

##### ANSIBLE VERSION

1.9.2

##### SUMMARY

From @giacmir on 2015-10-08T13:11:51Z
I have a deploy scenario where I have worktree and git-dir separated. if I try to update using the git module it fails with this error

```
FAILED >> {
    "cmd": "/usr/bin/git clone --origin origin --branch master '' /path/to/worktree", 
    "failed": true, 
    "msg": "fatal: destination path '/path/to/worktree' already exists and is not an empty directory.", 
    "rc": 128, 
    "stderr": "fatal: destination path '//path/to/worktree' already exists and is not an empty directory.\n", 
    "stdout": ""
}
```

Same happens if I put the git dir as dest.
Ansible is version 1.9.2 and git 1.7.1

Copied from original issue: ansible/ansible-modules-core#2226

### PR DESCRIPTION
**Pull request by cmrd-senya**
##### SUMMARY

fixes #30034 

Detects a separate git tree configuration in the working tree  and sets `$GIT_DIR` environment variable to that value

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

git.py

##### ANSIBLE VERSION

```
ansible 2.0.0.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

### CODE REVIEW

File: lib/ansible/modules/source_control/git.py [hunk 1]
```diff
@@ -1012,13 +1012,22 @@ def main():
     # call run_command()
     module.run_command_environ_update = dict(LANG='C', LC_ALL='C', LC_MESSAGES='C', LC_CTYPE='C')
 
+    separate_git_dir = None
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: There's no point of having this var here unless you're going to reuse it after the checks. Please remove this line.

**Comment by cmrd-senya** [comment]
I updated the PR. There is a failing test: https://app.shippable.com/github/ansible/ansible/runs/59445/27/tests

Looks like it's not my fault:
```
fatal: unable to access 'https://github.com/jimi-c/test_role/': Failed to connect to github.com port 443: Operation timed out
```

**Comment by mattclay** [comment]
I've restarted CI for this PR.

File: lib/ansible/modules/source_control/git.py [hunk 2]
```diff
@@ -1019,6 +1019,14 @@ def main():
         dest = os.path.abspath(dest)
         if bare:
             gitconfig = os.path.join(dest, 'config')
+        elif os.path.isfile(os.path.join(dest, '.git')):
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+                match = re.match('gitdir: (.+)$', data)
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: Please dedent all the following code, so that it would be outside of `with`-block.

**Comment by cmrd-senya** [comment]
@webknjaz, updated

**Comment by webknjaz** [comment]
@cmrd-senya could you please add tests for this?
https://github.com/ansible/ansible/tree/devel/test/integration/targets/git/tasks

**Comment by cmrd-senya** [comment]
@webknjaz, updated

File: lib/ansible/modules/source_control/git.py [hunk 3]
```diff
@@ -1019,6 +1019,14 @@ def main():
         dest = os.path.abspath(dest)
         if bare:
             gitconfig = os.path.join(dest, 'config')
+        elif os.path.isfile(os.path.join(dest, '.git')):
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+            match = re.match('gitdir: (.+)$', data)
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: should the line start with `gitdir` or should it be in the end of line with smth prefixed? you might want to prefix the regex with `^`.

On the other hand, taking into account that regexps are expensive, you could just use str.split or partition with bits of validation on top, like:
```python
try:
    ref_prefix, separate_git_dir = data.split('gitdir: ', 1)
    if ref_prefix:
        raise ValueError('.git file has invalid git dir reference format')
    if not os.path.isdir(separate_git_dir):
        raise TypeError('%s is not a directory' % separate_git_dir)
    os.environ['GIT_DIR'] = separate_git_dir
except (TypeError, ValueError) as err:
    """``.git`` file does not have a valid format for detached Git dir."""
    module.fail_json(
        msg='Current repo does not have a valid reference to a '
        'separate Git dir or it refers to the invalid path',
        details=str(err),
    )
```

File: lib/ansible/modules/source_control/git.py [hunk 4]
```diff
@@ -1019,6 +1019,14 @@ def main():
...
+            if match:
+                separate_git_dir = match.group(1)
+                os.environ['GIT_DIR'] = separate_git_dir
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: I would not assign it w/o validation:
1) it might cause errors happening in some other place, which would be hard to spot
2) it might be a weak place, vulnerable to injection attacks (security issue)

File: test/integration/targets/git/tasks/main.yml [hunk 5]
```diff
@@ -36,3 +36,8 @@
 - include_tasks: reset-origin.yml
 - include_tasks: ambiguous-ref.yml
 - include_tasks: archive.yml
+- include_tasks: separate-git-dir.yml
+  when:
+    - not gpg_version.stderr
```
Comment (webknjaz) [inline test/integration/targets/git/tasks/main.yml]: I don't think it has anything to do with GPG commit signing.

File: test/integration/targets/git/tasks/separate-git-dir.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+# see https://github.com/ansible/ansible/pull/38016
+# see https://github.com/ansible/ansible/issues/30034
+
+- name: SEPARATE-GIT-DIR | clear checkout_dir
+  file:
+    state: absent
+    path: '{{ checkout_dir }}'
+
+- name: SEPARATE-GIT-DIR | clone with a separate git dir
+  command: git clone {{ repo_format1 }} git --separate-git-dir=separate.git
```
Comment (webknjaz) [inline test/integration/targets/git/tasks/separate-git-dir.yml]: For safety reasons, I'd create some random dir using `mktemp -d`

File: test/integration/targets/git/tasks/separate-git-dir.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  args:
+    chdir: "{{ output_dir }}"
+
+- name: SEPARATE-GIT-DIR | update repo the usual way
+  git:
+    repo: "{{ repo_format1 }}"
+    dest: "{{ checkout_dir }}"
```
Comment (webknjaz) [inline test/integration/targets/git/tasks/separate-git-dir.yml:47]: Also don't forget to clean-up dirs after testing (tear down stage)

File: test/integration/targets/git/tasks/separate-git-dir.yml (hunk 6 above)
Comment (cmrd-senya) [reply test/integration/targets/git/tasks/separate-git-dir.yml]: A random dir for `separate.git` directory? What kind of safety do you mean? If the problem is that it may already exist then we can clear the directory the same way as the `checkout_dir`.

File: lib/ansible/modules/source_control/git.py (hunk 3 above)
Comment (cmrd-senya) [reply lib/ansible/modules/source_control/git.py]: Do I have to write test cases for validations then?

Comment (cmrd-senya) [reply lib/ansible/modules/source_control/git.py]: Also, what is the better way to handle the exceptions: to keep trying the execution or to shut down immediately?

Comment (webknjaz) [reply lib/ansible/modules/source_control/git.py]: Fail fast and loud for fatal errors like being unable to fulfill the module's purpose. I've updated a snippet in the comment above to reflect this.
And yes, negative test cases for error code paths are highly appreciated, please add them.

File: test/integration/targets/git/tasks/separate-git-dir.yml (hunk 6 above)
Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml]: yes, i'm concerned about it already existing in place, but also current approach will cause conflicts and false failures in case of simultaneous test runs.

File: lib/ansible/modules/source_control/git.py [hunk 8]
```diff
@@ -1019,6 +1019,14 @@ def main():
         dest = os.path.abspath(dest)
         if bare:
             gitconfig = os.path.join(dest, 'config')
+        elif os.path.isfile(os.path.join(dest, '.git')):
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+            match = re.match('gitdir: (.+)$', data)
+            if match:
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: Also, in case of not match `gitconfig` would have `None` value, which would've broken this module in some place being executed later.

File: lib/ansible/modules/source_control/git.py (hunk 4 above)
Comment (webknjaz) [reply lib/ansible/modules/source_control/git.py]: @cmrd-senya I've got concerns about modifying globally accessible env vars, so after talking to a colleague he brought up that whenever you change the env var value you should revert it to its initial state after using it.
So I strongly recommend moving to CLI arg instead of messing with env vars.

File: lib/ansible/modules/source_control/git.py (hunk 3 above)
Comment (webknjaz) [reply lib/ansible/modules/source_control/git.py]: @cmrd-senya I've updated your branch with my implementation for now.

File: test/integration/targets/git/tasks/separate-git-dir.yml (hunk 6 above)
Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml]: Why not use `{{ checkout_dir }}` instead of `git` arg + chdir?

File: lib/ansible/modules/source_control/git.py [hunk 9]
```diff
@@ -1019,6 +1019,23 @@ def main():
         dest = os.path.abspath(dest)
         if bare:
             gitconfig = os.path.join(dest, 'config')
+        elif os.path.isfile(os.path.join(dest, '.git')):
+            with open(os.path.join(dest, '.git'), 'r') as gitfile:
+                data = gitfile.read()
+            try:
+                ref_prefix, os.environ['GIT_DIR'] = data.split('gitdir: ', 1)
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: Honestly, I don't understand why you'd want to change this env var at all. If the repo already has a separate git dir it will figure it out automatically. The only step git looks at this var is clone/init phase, for which there's an appropriate `--separate-git-dir` CLI argument, which is preferable over env var.

Comment (cmrd-senya) [reply lib/ansible/modules/source_control/git.py]: Because in order to set `--separate-git-dir` I would have to modify every function call and definition in this file which calls a git command

**Review by webknjaz — CHANGES_REQUESTED** [review]
So to summarize comments:
1) we need to support `separate-git-dir` module argument
2) have docs/examples for that
3) _probably_ drop setting an env var
4) cover all of that with tests including negative ones

File: lib/ansible/modules/source_control/git.py (hunk 9 above)
Comment (webknjaz) [reply lib/ansible/modules/source_control/git.py]: This place in code **does not affect** cloning/initing, while operating on **existing repo**, which makes it pointless. Also, it would be architecturally wrong to do it from some random phase of module execution. We would still have to bypass it to clone/init anyway.

File: test/integration/targets/git/tasks/separate-git-dir.yml (hunk 6 above)
Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml]: After adding support for module arg, you'll need to just use the git module itself rather then shell command.

File: lib/ansible/modules/source_control/git.py (hunk 9 above)
Comment (cmrd-senya) [reply lib/ansible/modules/source_control/git.py]: Why? The problem at #30034 is related to existing repo only.

Comment (cmrd-senya) [reply lib/ansible/modules/source_control/git.py]: The point of this PR is not to support cloning/initing with a separate git dir, the point is to support updating the existing repositories with separate git dir.

**Comment by webknjaz** [comment]
You'll need to bypass an arg into `clone` function and also I've noticed some similar config lookup in `get_head_branch` function, so it might also need changes.

Besides above, you'll need to tests this combined with submodules, since I've got a feeling that they might get broken with such changes. It seems like setting `GIT_DIR` var might be already breaking that.

File: lib/ansible/modules/source_control/git.py (hunk 9 above)
Comment (webknjaz) [reply lib/ansible/modules/source_control/git.py]: 1) It does make almost no sense to support this feature of Git only partially.
2) `GIT_DIR` env var does not affect fetch, since Git would already parse `gitdir: ...` from `.git` file to figure out where the repo is located.
3) Also, the very first snippet (and the issue description) in #30034 showcases `clone` happening there. Which happens because of `not os.path.exists(gitconfig)`, not because of env var.

**Comment by cmrd-senya** [comment]
> we need to support separate-git-dir module argument

I disagree. My intention is to fix #30034. This PR handles the case where the existing repo with separate git dir is updated, which can be detected automatically by examining `.git` type and contents and therefore doesn't require any additional arguments. I'm currently not interested in implementing support for cloning new repositories with separate git dirs in ansible.

**Comment by webknjaz** [comment]
Okay, I can agree with not adding a new module argument in this PR, but I don't agree to accept it without proper testing and with env var modification (which is not needed).

Will be waiting for those then and will do the next round of review afterwards.

File: lib/ansible/modules/source_control/git.py (hunk 9 above)
Comment (cmrd-senya) [reply lib/ansible/modules/source_control/git.py]: > Also, the very first snippet (and the issue description) in #30034 showcases clone happening there. Which happens because of not os.path.exists(gitconfig), not because of env var.

Clone is happening right because the ansible git module fails to detect the existing cloned repository. If you read the issue description carefully it says:

>  if I try to update using the git module it fails with this error

**Comment by cmrd-senya** [comment]
> but I don't agree to accept it without proper testing and with env var modification (which is not needed).

Does it mean I have to change every function which calls git commands? Will I have to change every such function's argument list and pass separate git dir value down stream everywhere? Or is there a way to store the separate git dir value globally for this module?

**Comment by webknjaz** [comment]
I've just looked through `get_head_branch` function and it repeats part of logic for pulling out the gitdir from config, I suggest that it could be moved to a standalone function and reused to avoid further code diversity in future.

**Comment by webknjaz** [comment]
> Does it mean I have to change every function which calls git commands? Will I have to change every such function's argument list and pass separate git dir value down stream everywhere? Or is there a way to store the separate git dir value globally for this module?

No, your guess about role of env var was misleading, we just need to drop it, no other modifications needed.

File: lib/ansible/modules/source_control/git.py (hunk 9 above)
Comment (webknjaz) [reply lib/ansible/modules/source_control/git.py]: Yes, but it is still unrelated to the env var.

**Comment by cmrd-senya** [comment]
> No, your guess about role of env var was misleading, we just need to drop it, no other modifications needed.

Oh, right, it seems to work without it :+1:

**Comment by webknjaz** [comment]
Yes, as I told it only affects clone/init, which was very confusing to me and I had to research it myself first.

**Comment by cmrd-senya** [comment]
Sorry about that. I was mistakenly sure that this env var is require to properly handle existing git repositories, while apparently the `.git` does this job in this case.

**Comment by mattclay** [comment]
CI failure in integration tests `test/integration/targets/git/tasks/separate-git-dir.yml:15`:

> failure: Current repo does not have a valid reference to a separate Git dir or it refers to the invalid path

**Comment by cmrd-senya** [comment]
Updated, please review

File: test/integration/targets/git/tasks/separate-git-dir.yml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+- name: SEPARATE-GIT-DIR | clone with a separate git dir
+  command: git clone {{ repo_format1 }} {{ checkout_dir }} --separate-git-dir={{ tempdir.stdout }}
+
+- name: SEPARATE-GIT-DIR | update repo the usual way
+  git:
+    repo: "{{ repo_format1 }}"
+    dest: "{{ checkout_dir }}"
+
+- name: SEPARATE-GIT-DIR | set git dir to non-existent dir
+  shell: "echo gitdir: /dev/null/non-existent-dir > .git"
```
Comment (webknjaz) [inline test/integration/targets/git/tasks/separate-git-dir.yml:23]: We also need to test the case where `gitdir: ...` is on the other line.

Comment (cmrd-senya) [reply test/integration/targets/git/tasks/separate-git-dir.yml:23]: Like this https://github.com/ansible/ansible/pull/38016/commits/82430f6c2d0c80475177cbb2dbae0857606abeeb ?

File: test/integration/targets/git/tasks/separate-git-dir.yml [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  git:
+    repo: "{{ repo_format1 }}"
+    dest: "{{ checkout_dir }}"
+  ignore_errors: yes
+  register: result
+
+- name: SEPARATE-GIT-DIR | check update has failed
+  assert:
+    that:
+      - result is failed
```
Comment (webknjaz) [inline test/integration/targets/git/tasks/separate-git-dir.yml:37]: it should not fails if `.gitdir` contains a list of `key: value`

File: test/integration/targets/git/tasks/separate-git-dir.yml (hunk 10 above)
Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml:23]: Except it should succeed

Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml:23]: That's why I asked you to reuse existing code: https://github.com/ansible/ansible/pull/38016#issuecomment-387800545

Comment (cmrd-senya) [reply test/integration/targets/git/tasks/separate-git-dir.yml:23]: Okay

Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml:23]: To be clear: don't copy-paste blindly, merge the best parts of the approaches ;)

File: test/integration/targets/git/tasks/separate-git-dir.yml (hunk 11 above)
Comment (cmrd-senya) [reply test/integration/targets/git/tasks/separate-git-dir.yml:37]: I can't find any reference that proves that. Git documentation [says](https://git-scm.com/docs/gitrepository-layout#_description):
>  Also you can have a plain text file .git at the root of your working tree, containing gitdir: <path> to point at the real directory that has the repository.

And no information about the key-value list.

I also tried a few examples, including
```
key: value
gitdir: ...path
```
and (with newline at the front):
```

gitdir: ...path
```

And all of the examples fail at the repository update stage with the message from git: `fatal: Invalid gitfile format: .git`. Can you please provide any correct examples of key-value lists for the `.git` file which will work with Git?

Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml:37]: Not really, but since that other snippet loops through key-value lines there must be something working this way. I'll try investigating this.

Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml:37]: It looks like the best way of checking integrity of the repo is [`git-fsck`](https://git-scm.com/docs/git-fsck), but I don't thing that it's needed here now.

Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml:37]: But you are right, it does contain a validation for a file immediately starting with `gitdir: `: https://github.com/git/git/blob/master/setup.c#L592

Comment (webknjaz) [reply test/integration/targets/git/tasks/separate-git-dir.yml:37]: So now we know that it should only contain `gitdir: ...path` and nothing more in that file. Trailing EOLs are being stripped off by git.

**Review by webknjaz — APPROVED** [review]

**Comment by cmrd-senya** [comment]
But don't you think we still need to deduplicate the code with `get_head_branch`?

**Comment by webknjaz** [comment]
Right, that's why I haven't merged it yet.

**Comment by cmrd-senya** [comment]
I was confused because you approved the PR. I already made the changes locally, but I need time to fix them according to our discoveries about `.git` file. I'll do that tomorrow.

**Comment by cmrd-senya** [comment]
Updated

File: lib/ansible/modules/source_control/git.py [hunk 12]
```diff
@@ -595,14 +595,7 @@ def is_not_a_branch(git_path, module, dest):
...
 
-def get_head_branch(git_path, module, dest, remote, bare=False):
-    '''
-    Determine what branch HEAD is associated with.  This is partly
-    taken from lib/ansible/utils/__init__.py.  It finds the correct
-    path to .git/HEAD and reads from that file the branch that HEAD is
-    associated with.  In the case of a detached HEAD, this will look
-    up the branch in .git/refs/remotes/<remote>/HEAD.
-    '''
+def get_repo_path(module, dest, bare):
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: Returning an error on Ansible module protocol level shouldn't be mixed up with just identifying a directory.
Use exceptions to communicate failures to the calling code and process them there, outside of this function.
So no `module` argument is needed here — let's separate different abstraction layers.

File: lib/ansible/modules/source_control/git.py [hunk 13]
```diff
@@ -611,23 +604,39 @@ def get_head_branch(git_path, module, dest, remote, bare=False):
     if os.path.isfile(repo_path):
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: Now that we support separate git dir the comment above requires extension.

File: lib/ansible/modules/source_control/git.py [hunk 14]
```diff
@@ -611,23 +604,39 @@ def get_head_branch(git_path, module, dest, remote, bare=False):
...
         try:
             git_conf = open(repo_path, 'rb')
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: This is a forgotten, unclosed and unnecessary file descriptor. We need to wipe this line.

File: lib/ansible/modules/source_control/git.py (hunk 12 above)
Comment (cmrd-senya) [reply lib/ansible/modules/source_control/git.py]: BTW, it feels like calling `get_repo_path` for the second time for the same `dest` is somehow redundant. `get_repo_path` is called in `main` anyway. Is there a way to save the result after the first invocation and use it later?

Comment (webknjaz) [reply lib/ansible/modules/source_control/git.py]: Yes, it's called [`lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache), but it's not available in stdlib of Python 2.7, so we cannot use it (unless we vendor [it's backport](https://pypi.org/p/backports.functools_lru_cache) into the project like `six` distribution).

File: lib/ansible/modules/source_control/git.py [hunk 15]
```diff
@@ -611,23 +604,39 @@ def get_head_branch(git_path, module, dest, remote, bare=False):
...
 
             # There is a possibility the .git file to have an absolute path.
             if os.path.isabs(gitdir):
                 repo_path = gitdir
             else:
                 repo_path = os.path.join(repo_path.split('.git')[0], gitdir)
-        except (IOError, AttributeError):
+            if not os.path.isdir(repo_path):
+                raise ValueError('%s is not a directory' % repo_path)
+        except (IOError, AttributeError, ValueError) as err:
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py]: Why `IOError` or `AttributeError` can happen? I can see only `ValueError` occuring.

File: lib/ansible/modules/source_control/git.py [hunk 16]
```diff
@@ -603,31 +626,16 @@ def get_head_branch(git_path, module, dest, remote, bare=False):
...
-            if os.path.isabs(gitdir):
-                repo_path = gitdir
-            else:
-                repo_path = os.path.join(repo_path.split('.git')[0], gitdir)
-        except (IOError, AttributeError):
-            # No repo path found
-            return ''
+    try:
+        repo_path = get_repo_path(dest, bare)
+    except (IOError, ValueError) as err:
```
Comment (webknjaz) [inline lib/ansible/modules/source_control/git.py:632]: @cmrd-senya `IOError` can happen in case of insufficient privileges or similar issues while accessing fs, but I cannot see `AttributeError` appearing. Do you recall anything causing it?

Comment (cmrd-senya) [reply lib/ansible/modules/source_control/git.py:632]: I don't. `AttributeError` was present in the original code and I don't think I ever saw it being thrown.

**Comment by cmrd-senya** [comment]
Updated

**Comment by cmrd-senya** [comment]
@webknjaz, thank you for your assistance!

File: test/integration/targets/git/tasks/main.yml [hunk 17]
```diff
@@ -36,3 +36,6 @@
 - include_tasks: reset-origin.yml
 - include_tasks: ambiguous-ref.yml
 - include_tasks: archive.yml
+- include_tasks: separate-git-dir.yml
+  when:
+    - git_version.stdout is version("1.7.5", '>=')
```
Comment (webknjaz) [inline test/integration/targets/git/tasks/main.yml:41]: I missed that it should've been `git_version.stdout | version_compare("1.7.5", '>=')`

Comment (cmrd-senya) [reply test/integration/targets/git/tasks/main.yml:41]: I copied that from the line 34
